contracts/peripheral/grants/VeBalGrant.sol

```diff
@@ -0,0 +1,352 @@
+     */
+    modifier onlyProject() {
+        require(msg.sender == project, "!project");
+        _;
+    }
+
+    /**
+     * @notice Modifier that allows only Balancer to trigger a function
+     */
+    modifier onlyBalancer() {
+        require(msg.sender == balancer, "!balancer");
+        _;
+    }
+
+    /**
+     * @notice Modifier that only allows something to be called when the contract is active
+     */
+    modifier whileActive() {
+        require(active, "!active");
+        _;
+    }
+
+    /**
+     * @notice Modifier that only allows something to be called when the contract is inactive
+     */
+    modifier whileInactive() {
+        require(!active, "active");
+        _;
+    }
+
+    /* ----------------------------------------------------------------
+       Shared Functions
+    ---------------------------------------------------------------- */
+
+    /**
+     * @notice Releases veBAL lock
+     * grant must be inactive in order for this to be called
+     */
+    function release() external onlyAuth whileInactive {
+        votingEscrow.withdraw();
+        hasLock = false;
+    }
+
+    /**
+     * @notice exits BPT position in return for WETH and BAL
+     * grant must be inactive in order for this to be called
+     */
+    function redeem() external onlyAuth whileInactive {
+        _exitBalEthPool();
+    }
+
+    /**
+     * @notice Sends WETH and BAL to project and balancer
+     * Project gets WETH up to the amount they contributed in the initial lock
+     * Balancer gets all BAL and any remaining WETH
+     * grant must be inactive in order for this to be called
+     */
+    function withdrawBalances() external onlyAuth whileInactive {
+        uint256 wethForProjectBalance = AuraMath.min(ethContributed, WETH.balanceOf(address(this)));
+        WETH.transfer(project, wethForProjectBalance);
+        WETH.transfer(balancer, WETH.balanceOf(address(this)));
+        BAL.transfer(balancer, BAL.balanceOf(address(this)));
+        ethContributed = 0;
+    }
+
+    /**
+     * @notice Allows Balancer or Project to vote for a gauge
+     * @param gauge      gauge that will be voted for
+     * @param weight     vote weight
+     */
+    function voteGaugeWeight(address gauge, uint256 weight) external {
+        active ? require(msg.sender == project, "!caller") : require(msg.sender == balancer, "!caller");
+        gaugeController.vote_for_gauge_weights(gauge, weight);
+    }
+
+    /**
+     * @notice Allows Balancer or Project to call other contracts via the grant
+     * @notice some addresses and function selectors are barred from being called
+     * @param _to      Target Contract
+     * @param _value   Eth to be sent
+     * @param _data    Call data
+     */
+    function execute(
+        address _to,
+        uint256 _value,
+        bytes memory _data
+    ) external returns (bool, bytes memory) {
+        active ? require(msg.sender == project, "!caller") : require(msg.sender == balancer, "!caller");
```

Comment:
Should we also restrict deposits into gauges? Considering the agreed use of the veBAL votes is to vote for their ve8020 token.

Reply:
Maybe it doesn't matter actually as both `balancer` and `project` have access to this function.

Reply:
Yeah balancer can cancel the grant if a project goes rogue with it I guess